contracts/transfer/src/state.rs

```diff
@@ -223,28 +228,164 @@ impl TransferState {
+            self.gas_price = None;
+            if let Ok(RawResult {
+                data: _,
+                economic_mode,
+            }) = result.clone()
+            {
+                match economic_mode {
+                    EconomicMode::Charge(charge) if charge != 0 => {
+                        rusk_abi::set_charge(charge)
+                    }
+                    EconomicMode::Allowance(allowance) if allowance != 0 => {
+                        rusk_abi::set_allowance(allowance)
+                    }
+                    _ => (),
+                }
+            }
+        }
+
+        result.map(|r| r.data)
+    }
+
+    // Applies contract's charge. Caller of the contract will pay a
+    // larger fee so that contract can earn the difference between
+    // charge and the actual cost of the call.
+    // Charge has no effect if the actual cost of the call is greater
+    // than the charge.
+    // Returns economic gas spent
+    fn apply_charge(
+        &mut self,
+        contract_id: &ContractId,
+        charge: u64,
+        gas_spent: u64,
+        gas_price: u64,
+    ) -> u64 {
+        let cost = gas_spent * gas_price;
+        if charge > cost {
+            let earning = charge * gas_price - cost;
+            self.add_balance(*contract_id, earning);
+            rusk_abi::emit(
+                "earning",
+                EconomicEvent {
+                    module: contract_id.to_bytes(),
+                    value: earning,
+                    result: EconomicResult::ChargeApplied,
+                },
             );
+            charge
+        } else {
+            rusk_abi::emit(
+                "earning",
+                EconomicEvent {
+                    module: contract_id.to_bytes(),
+                    value: charge * gas_price,
+                    result: EconomicResult::ChargeNotSufficient,
+                },
+            );
+            gas_spent
         }
+    }
 
-        result
+    // Applies contract's allowance. Caller of the contract's method
+    // won't pay a fee and all the cost will be covered by the contract.
+    // Allowance has no effect if contract does not have enough funds or
+    // if the actual cost of the call is greater than allowance.
+    // Returns economic gas spent
+    fn apply_allowance(
+        &mut self,
+        contract_id: &ContractId,
+        allowance: u64,
+        gas_spent: u64,
+        gas_price: u64,
+    ) -> u64 {
+        let spent = gas_spent * gas_price;
+        if allowance * gas_price < spent {
+            rusk_abi::emit(
+                "sponsoring",
+                EconomicEvent {
+                    module: contract_id.to_bytes(),
+                    value: allowance * gas_price,
+                    result: EconomicResult::AllowanceNotSufficient,
+                },
+            );
+            gas_spent
+        } else {
+            let contract_balance = self.balance(contract_id);
```

Comment:
I believe we have a sort of confusion here.

This "balance" is just the amount of the notes that have been sent to that contract. You cannot deduct from there.

Let's take the stake contract as example.
Everyone that a user is performing a stake, a STCT is performed, increasing the balance.
That balance is then checked when performing WFCT while unstaking, to avoid mint/burn of funds...

But keep in mind, that having value in that balance doesn't means that those funds are available to the owner of the contract to "sponsor" os "subsidize" or whatever you want.

I strongly believe we should make a differentiation between contract "balance" and contract "wallet"

Reply:
I am not sure I agree. Contract balance is the amount of notes that have been "burned" into it. Any withdrawal from this balance is only possible if you mint new notes or pay fees. So the overall value balance is preserved. As long as the overall "notes worlds value balance" is preserved, everything is fine IMO. Let's not confuse it with the stake contract which has completely different logic.

Reply:
Of course question remains, whose are the funds which are in contract's balance. I think it is a healthy assumption that the funds belong to the contract. If in case of stake contract this is not the case, that may be a problem, as theoretically stake contract could use these funds to pay for fees etc. But stake contract is something we control, it is a genesis contract, so we can ensure that it will behave and not use users' funds inappropriately. The same logic applies to any other contract - it is responsible for its funds. We could think of a separate wallet for contracts - question is, how different would it be from what we have now.